janitor/functions/complete.py

```diff
@@ -212,112 +212,141 @@ def _computations_complete(
+                    # for categorical dtypes, set the categories first
+                    if is_categorical_dtype(df[column_name]):
+                        df[column_name] = df[column_name].cat.add_categories(
+                            [value]
+                        )
+                    df.loc[boolean_filter, column_name] = value
+
+    if not df.columns.equals(columns):
+        return df.reindex(columns=columns)
     return df
 
 
-def _generic_complete(df: pd.DataFrame, columns: list, all_strings: bool):
+def _generic_complete(
+    df: pd.DataFrame, columns: list, all_strings: bool, sort: bool
+):
     """
     Generate cartesian product for `_computations_complete`.
 
-    Returns a Series or DataFrame, with no duplicates.
+    Returns a DataFrame, with no duplicates.
     """
     if all_strings:
-        uniques = {col: df[col].unique() for col in columns}
+        if sort:
+            uniques = {}
+            for col in columns:
+                column = extract_array(df[col], extract_numpy=True)
+                _, column = pd.factorize(column, sort=sort)
+                uniques[col] = column
+        else:
+            uniques = {col: df[col].unique() for col in columns}
         uniques = _computations_expand_grid(uniques)
-        uniques = uniques.droplevel(level=-1, axis="columns")
+        uniques.columns = columns
         return uniques
 
     uniques = {}
+    df_columns = []
     for index, column in enumerate(columns):
+        if not isinstance(column, str):
+            df_columns.extend(column)
+        else:
+            df_columns.append(column)
         if isinstance(column, dict):
-            column = _complete_column(column, df)
+            column = _complete_column(column, df, sort)
             uniques = {**uniques, **column}
         else:
-            uniques[index] = _complete_column(column, df)
+            uniques[index] = _complete_column(column, df, sort)
 
     if len(uniques) == 1:
         _, uniques = uniques.popitem()
         return uniques.to_frame()
 
     uniques = _computations_expand_grid(uniques)
-    return uniques.droplevel(level=0, axis="columns")
+    uniques.columns = df_columns
+    return uniques
 
 
 @functools.singledispatch
-def _complete_column(column: str, df):
+def _complete_column(column: str, df, sort):
     """
     Args:
         column : str/list/dict
         df: Pandas DataFrame
 
     A Pandas Series/DataFrame with no duplicates,
-    or a list of unique Pandas Series is returned.
+    or a dictionary of unique Pandas Series is returned.
     """
+    # the cost of checking uniqueness is expensive,
+    # especially for large data
+    # dirty tests also show that drop_duplicates
+    # is faster than pd.unique for fairly large data
+
     column = df[column]
+    dupes = column.duplicated()
+
+    if dupes.any():
+        column = column[~dupes]
+
+    if sort and (column.is_monotonic_increasing is False):
```

Comment:
They do d same thing, and I find the first one more explicit (I am guilty of using them interchangeably tho)

Reply:
Yeah, they do.
The second is more widely regarded to be pythonic, though. (`is False` is really only used if there is an explicit need to check for False rather than generic Falsy values like 0, None, `[]`, etc.)

<img width="819" alt="image" src="https://user-images.githubusercontent.com/30731072/156930565-825f6eac-abe4-4cbc-9573-538508ac8d79.png">
https://www.python.org/dev/peps/pep-0008/